[GEP-26] Use WorkloadIdentity for ETCD backup - part 3

vpnachev opened the pull request:
**How to categorize this PR?**
/area security ipcei
/kind enhancement
/label ipcei/workload-identity 

**What this PR does / why we need it**:
Enables gardenlet.backupbucket, gardenlet.backupentry, and extensions.backupentry controllers with support for credentials of type WorkloadIdentity  

**Which issue(s) this PR fixes**:
Part of #9586 

**Special notes for your reviewer**:

**Release note**:
<!--
Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       breaking|noteworthy|feature|bugfix|doc|other
- target_group:   user|operator|developer|dependency
-->
```feature developer
BackupBucket/BackupEntry controllers now support WorkloadIdentity type of credentials, provider extensions may need to adjust the respective controllers or to explicitly disallow BackupBuckets of their type to configure WorkloadIdentity. 
```

vpnachev committed:
Some minor changes and improvements

vpnachev committed:
Prepare util func GetCredentialsByObjectReference

vpnachev committed:
Add support for BackupBucket as context of workload identity token

vpnachev committed:
Add method to compare generated workload identity secret

vpnachev committed:
Add support for WorkloadIdentity in gardenlet backupbucket controller

vpnachev committed:
Do not read twice backupbucket credentials

vpnachev committed:
Add support for WorkloadIdentity in gardenlet backupentry controller

vpnachev committed:
Add support for BackupEntry as context of workload identity token

vpnachev committed:
Create informers for workloadidentity storage before the SharedInformerFactory is started

This actually fixes an issue where the gardener-apiserver could not list
an existing BackupEntry during issuance of workload identity token

vpnachev committed:
Add support for WorkloadIdentity in extensions backupentry controller

vpnachev committed:
Flatten a bit conditional expressions in extensions backupentry mapper

vpnachev committed:
Unify string concatenation in extensions backupentry util

vpnachev committed:
Update skaffold deps

dimityrmirchev commented on extensions/pkg/controller/backupentry/genericactuator/actuator.go:
```diff
@@ -54,33 +59,80 @@ func (a *actuator) deployEtcdBackupSecret(ctx context.Context, log logr.Logger,
+
+	if purpose, ok := backupEntrySecret.Labels[securityv1alpha1constants.LabelPurpose]; ok && purpose == securityv1alpha1constants.LabelPurposeWorkloadIdentityTokenRequestor {
+		withWorkloadIdentity = true
+	}
```
I think this is simpler to read.
```suggestion
	var (
		backupEntrySecretData = backupEntrySecret.DeepCopy().Data
		useWorkloadIdentity = backupEntrySecret.Labels[securityv1alpha1constants.LabelPurpose] == securityv1alpha1constants.LabelPurposeWorkloadIdentityTokenRequestor
	)
```

vpnachev replied:
Done with a5390c9350b745c138fb26b92cef96b9f1fc70c8

dimityrmirchev commented on extensions/pkg/controller/backupentry/genericactuator/actuator.go:
```diff
@@ -54,33 +59,80 @@ func (a *actuator) deployEtcdBackupSecret(ctx context.Context, log logr.Logger,
+	}
+
+	if withWorkloadIdentity {
+		maps.DeleteFunc(backupEntrySecretData, func(k string, _ []byte) bool {
+			return k != securityv1alpha1constants.DataKeyConfig
+		})
```
Just for my understanding: Why are we removing other keys from the secret? Do we expect anything else to be set here? Maybe we can leave an explaining comment.

vpnachev replied:
As per https://github.com/gardener/gardener/blob/master/docs/proposals/26-workload-identity.md, the only standard data keys are `token` and `config`, all others should be set by admission controllers. 
By removing them here we ensure that we are not transporting data from the backup entry secret to the etcd-backup secret that could be considered by the admission controller as already set and should not be touched. 

And the `token` is removed to let the token requestor controller generate fresh token for etcd-backup secret instead of re-using the one from the backup entry.

vpnachev replied:
Done with bc3da1a943a6b4d3e7cfa41c0b6284c92b5e3537

dimityrmirchev commented on extensions/pkg/controller/backupentry/genericactuator/actuator.go:
```diff
@@ -54,33 +59,80 @@ func (a *actuator) deployEtcdBackupSecret(ctx context.Context, log logr.Logger,
 
 	_, err = controllerutils.GetAndCreateOrMergePatch(ctx, a.client, etcdSecret, func() error {
+		if withWorkloadIdentity {
+			maps.DeleteFunc(etcdSecret.Annotations, func(k, _ string) bool {
+				return k != securityv1alpha1constants.AnnotationWorkloadIdentityTokenRenewTimestamp
+			})
```
Same question as above.

vpnachev replied:
It is similarly to the data keys. If there are other annotations that needs to be present, they should be set via an admission controller. It is also aligned with the behavior of https://github.com/gardener/gardener/blob/master/pkg/utils/workloadidentity/secret.go.

vpnachev replied:
Done with bc3da1a943a6b4d3e7cfa41c0b6284c92b5e3537

dimityrmirchev commented on extensions/pkg/controller/backupentry/genericactuator/actuator.go:
```diff
@@ -54,33 +59,80 @@ func (a *actuator) deployEtcdBackupSecret(ctx context.Context, log logr.Logger,
+			}
+
+			if token, ok := etcdSecret.Data[securityv1alpha1constants.DataKeyToken]; ok {
+				etcdSecretData[securityv1alpha1constants.DataKeyToken] = token
```
Can we have a race condition if the token requestor updates the secret with a new token, but we send a patch request with the old one?

vpnachev replied:
Good question. I think there shouldn't be race condition, the intent here is just to preserve the existing token, if it is already set. This way the generated patch should not be concerning the `data.token` field and no change should be applied.

vpnachev replied:
Locally testing this is not overwriting the `token` value when it is concurrently updated by other controllers. But it may skip updating the `config` value and to reset it to the expected value. 

For consistency with the other components, we will use `OptimisticLocking` (ref: 4dbf242825e87e600982b8fd0bd350fc1f94cabc) here so that the patch will fail if it is not working with the latest version of the secret and re-try.

dimityrmirchev commented on pkg/apiserver/registry/security/workloadidentity/storage/jwt.go:
```diff
@@ -90,65 +93,139 @@ func (r *TokenRequestREST) getGardenerClaims(workloadIdentity *securityapi.Workl
+
 	default:
-		return nil, nil, nil, fmt.Errorf("unsupported GVK for context object: %s", gvk.String())
+		return nil, nil, nil, nil, nil, fmt.Errorf("unsupported GVK for context object: %s", gvk.String())
```
I suggest that we group these in a single struct as it will be easier to read.

vpnachev replied:
Sure, I will do it.

vpnachev replied:
Done with 14a023f22dadaf481bfa69cab155f7c7377e786a

dimityrmirchev commented on pkg/gardenlet/controller/backupbucket/reconciler.go:
```diff
@@ -105,17 +107,17 @@ func (r *Reconciler) reconcileBackupBucket(
-		if err := controllerutils.AddFinalizers(gardenCtx, r.GardenClient, gardenSecret, gardencorev1beta1.ExternalGardenerName); err != nil {
-			return fmt.Errorf("failed to add finalizer to backup secret: %w", err)
+	if !controllerutil.ContainsFinalizer(backupCredentials, gardencorev1beta1.ExternalGardenerName) {
+		log.Info("Adding finalizer to backup credentials", "kind", backupCredentials.GetObjectKind().GroupVersionKind().Kind, "key", client.ObjectKeyFromObject(backupCredentials))
```
Does it make sense to be more explicit here?
```suggestion
		log.Info("Adding finalizer to backup credentials", "gvk", backupCredentials.GetObjectKind().GroupVersionKind(), "key", client.ObjectKeyFromObject(backupCredentials))
```

vpnachev replied:
Actually, I did it this way at some point of time but I did not like how GVK is [stringified](https://github.com/kubernetes/kubernetes/blob/393ecc29a5add2f936805af21290a1960c04056d/staging/src/k8s.io/apimachinery/pkg/runtime/schema/group_version.go#L161-L163), but that is my personal perception and I am open to change it.  

```go
{"gvk": "security.gardener.cloud/v1alpha1, Kind=WorkloadIdentity", ....}
```

dimityrmirchev replied:
Yes, I also thought of that, but I still thought that it is worth the question if the "gvk" variant is better. Up to you, I am not against the other option as well since we only cover `Secret` and `WorkloadIdentity` and "kind" should be enough to identity them.

vpnachev replied:
Done with 8d92457565648459054806310f9527aac3f337e0

dimityrmirchev reviewed (changes requested):
Thanks for the well structured PR.

vpnachev committed:
Review feedback: Simplify workload identity usage detection

Co-authored-by: Dimitar Mirchev <dimitar.mirchev@sap.com>

vpnachev committed:
Refactor resolveContextObject to return single struct instead of multiple objects

Co-authored-by: Dimitar Mirchev <dimitar.mirchev@sap.com>

vpnachev committed:
Use OptimiscLocking when patching etcd-backup secret

This prevents concurrent updates by accident as token-requestor is also
writing to the same secret

Co-authored-by: Dimitar Mirchev <dimitar.mirchev@sap.com>

vpnachev committed:
Clarify why data keys, annotations and labels are reset in workload identity secret

dimityrmirchev commented on pkg/apiserver/registry/security/workloadidentity/storage/jwt.go:
```diff
@@ -68,87 +81,165 @@ func (r *TokenRequestREST) getGardenerClaims(workloadIdentity *securityapi.Workl
+		}
+
+		if backupBucket.UID != ctxObj.UID {
+			return nil, fmt.Errorf("uid of contextObject (%s) and real world resource(%s) differ", ctxObj.UID, backupBucket.UID)
```
Nit (also for other occurrences):
I wonder if we should make the error message more explicit about the type of objects. I imagine that only UIDs would be hard to find without additional context.

```suggestion
			return nil, fmt.Errorf("uid of contextObject (%s) and real world resource (%s) differ", ctxObj.UID, backupBucket.UID)
```

vpnachev replied:
The error is associated with the secret being reconciled, that should be enough to identify the context object. 
It is something like 
```json
{
    "Secret": {
        "name": "bucket-088001bb-6462-4212-bfdb-af0bb8f6db05",
        "namespace": "garden"
    },
    "namespace": "garden",
    "name": "bucket-088001bb-6462-4212-bfdb-af0bb8f6db05",
    "error": "could not request token for WorkloadIdentity: failed to resolve context object: uid of contextObject (c867e761-2b5a-4d7f-8af3-1b0ee7d2d3be) and real world resource(c867e761-2b5a-4d7f-8af3-1b0ee7d2d3bf) differ"
    ...,
}
```

As this error message is exactly the same for all context objects, I would just add the kind of object, i.e.


```suggestion
			return nil, fmt.Errorf("uid of contextObject (%s) and real world BackupBucket resource (%s) differ", ctxObj.UID, backupBucket.UID)
```

wdyt?

dimityrmirchev replied:
Yes, sound good

vpnachev replied:
Done with 1046ffcd0f7add618af9c1b76aaa7cef1ae90b39

dimityrmirchev reviewed (commented):
Thank you for addressing the feedback! I went over the PR one more time and managed to find a single nit.

/lgtm

vpnachev committed:
Improve error message

vpnachev committed:
Use GVK instead of Kind in log messages

shafeeqes commented on pkg/gardenlet/controller/backupbucket/reconciler_test.go:
```diff
@@ -166,100 +168,273 @@ var _ = Describe("Controller", func() {
+			extensionSecret.Type = "Opaque"
+		})
+
+		It("should create the extension secret and extension BackupBucket if it doesn't exist yet", func() {
```
Can we reduce the duplication by extracting the `It`s to a separate function? Same for the `backupentry` controller

vpnachev replied:
Can you elaborate what do you mean, which duplications are you concerned about and how would you propose to avoid them?

shafeeqes replied:
I meant `It`s like
```
It("should reconcile the extension BackupBucket if the secret update timestamp is after the extension last update time", func() {
```
```
It("should not reconcile the extension BackupBucket if the secret data or extension spec hasn't changed", func() {
```
```
It("should reconcile the extension BackupBucket if it's been 12hrs after the extension's last reconciliation", func() {
```
are repeated in both 
```
Describe("#Secrets Credentials", func() {
```
and
```
Describe("#WorkloadIdentity Credentials", func() {
```
So we can extract them and even others with some tweaks to a separate function and call it in both contexts.

vpnachev replied:
I've updated the tests and removed some test cases that are clearly duplications. please take a look again.

shafeeqes commented on pkg/gardenlet/controller/backupbucket/reconciler.go:
```diff
@@ -147,26 +149,39 @@ func (r *Reconciler) reconcileBackupBucket(
-	secretLastUpdateTime, err := time.Parse(time.RFC3339Nano, extensionSecret.Annotations[v1beta1constants.GardenerTimestamp])
-	if err != nil {
-		return err
+	secretLastUpdateTime := time.Time{}
+	if lastUpdateTime, ok := extensionSecret.Annotations[v1beta1constants.GardenerTimestamp]; ok {
+		if secretLastUpdateTime, err = time.Parse(time.RFC3339Nano, lastUpdateTime); err != nil {
```
Why do we need this change and the check `secretLastUpdateTime.IsZero()` later on? If the timestamp is not present, we reconcile the extension secret anyway before, so this annotation is always present, right? Same for the `backupentry` controller

vpnachev replied:
The controller does not re-read the secret after the call to `reconcileBackupBucketExtensionSecret`, it just happens that when the credentials are of type Secret, the variable `extensionSecret` is passed by reference and the current implementation adds the timestamp annotation to the same variable in-place. In this case the annotation is indeed always present. 

However, with credentials of type WorkloadIdentity, the annotation is set via new variable created by the `github.com/gardener/gardener/pkg/utils/workloadidentity.Secret.New()` method, this is why `extensionSecret` is not annotated with the timestamp annotation and the for `IsZero` is added.

shafeeqes replied:
Ah, okay!

vpnachev committed:
Fix unit tests

vpnachev committed:
Update backup(bucket|entry) tests when WorkloadIdentity credentials are used

shafeeqes commented on pkg/gardenlet/controller/backupbucket/reconciler_test.go:
```diff
@@ -218,7 +218,7 @@ var _ = Describe("Controller", func() {
 		})
 
-		It("should create the extension secret and extension BackupBucket if it doesn't exist yet", func() {
+		It("should create the extension secret and extension BackupBucket if then do not exist yet", func() {
```
Nit: typo `they`

shafeeqes commented on pkg/gardenlet/controller/backupbucket/reconciler_test.go:
```diff
@@ -166,100 +168,258 @@ var _ = Describe("Controller", func() {
+			}))
+		})
+
+		It("should not reconcile the extension BackupBucket if the secret data or extension spec hasn't changed", func() {
+			Expect(seedClient.Create(ctx, extensionSecret)).To(Succeed())
+			delete(extensionBackupBucket.Annotations, "gardener.cloud/operation")
+			Expect(seedClient.Create(ctx, extensionBackupBucket)).To(Succeed())
+
+			result, err := reconciler.Reconcile(ctx, request)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(result).To(Equal(reconcile.Result{}))
+
+			Expect(seedClient.Get(ctx, client.ObjectKeyFromObject(extensionBackupBucket), extensionBackupBucket)).To(Succeed())
+			Expect(extensionBackupBucket.Annotations).NotTo(HaveKey("gardener.cloud/operation"))
+		})
```
Can we drop this case as well?

vpnachev replied:
Let's keep it as it is.
While the test seems like duplicate, the secret is handled differently depending on the credentials type. Having in mind the other tests are not duplicates, I do not see much value of extracting just this one and parametrizing it on the credential type.

shafeeqes commented on pkg/gardenlet/controller/backupbucket/reconciler_test.go:
```diff
@@ -166,100 +168,258 @@ var _ = Describe("Controller", func() {
+				Name:       workloadIdentity.Name,
+			}
+			Expect(gardenClient.Create(ctx, backupBucket)).To(Succeed())
+			Expect(gardenClient.Create(ctx, workloadIdentity)).To(Succeed())
```
Nit:
```suggestion
			Expect(gardenClient.Create(ctx, workloadIdentity)).To(Succeed())

```

shafeeqes commented on pkg/gardenlet/controller/backupbucket/reconciler_test.go:
```diff
@@ -383,57 +393,32 @@ var _ = Describe("Controller", func() {
 			// the workloadIdentity.spec.targetSystem.providerConfig=nil will cause the `config` data key to be removed from the secret
 			extensionSecret.Data = map[string][]byte{"config": []byte("null")}
 			Expect(seedClient.Create(ctx, extensionSecret)).To(Succeed())
+			delete(extensionBackupBucket.Annotations, "gardener.cloud/operation")
```
Do we really need this?

vpnachev replied:
Not exactly, but it is nice to ensure that the annotation was not set in advance so that the test can prove the reconciliation sets it. I am changing it 
```go
Expect(extensionBackupBucket.Annotations).ToNot(HaveKey("gardener.cloud/operation"))
```

shafeeqes commented on pkg/gardenlet/controller/backupentry/reconciler_test.go:
```diff
@@ -186,83 +188,247 @@ var _ = Describe("Controller", func() {
+			extensionSecret.Data = gardenSecret.Data
+		})
+
+		It("should create the extension secret and extension BackupEntry if it doesn't exist yet", func() {
```
```suggestion
		It("should create the extension secret and extension BackupEntry if they do not exist yet", func() {
```

shafeeqes commented on pkg/gardenlet/controller/backupentry/reconciler_test.go:
```diff
@@ -186,83 +188,247 @@ var _ = Describe("Controller", func() {
+			}))
+		})
+
+		It("should not reconcile the extension BackupEntry if the secret data or extension spec hasn't changed", func() {
```
same here

vpnachev replied:
If you by "same here" mean it should be 
```suggestion
		It("should not reconcile the extension BackupEntry if the secret data or extension spec have not changed", func() {
```

I think it would be wrong. Or maybe you meant something else?

shafeeqes replied:
Meant the same as https://github.com/gardener/gardener/pull/12321#discussion_r2170609377

shafeeqes commented on pkg/gardenlet/controller/backupentry/reconciler_test.go:
```diff
@@ -186,83 +188,247 @@ var _ = Describe("Controller", func() {
+			extensionSecret.Type = "Opaque"
+		})
+
+		It("should create the extension secret and extension BackupEntry if it doesn't exist yet", func() {
```
Nit:
```suggestion
		It("should create the extension secret and extension BackupEntry if they do not exist yet", func() {
```

dimityrmirchev reviewed (approved):
/lgtm

plkokanov commented on pkg/gardenlet/controller/backupbucket/reconciler_test.go:
```diff
@@ -166,100 +168,258 @@ var _ = Describe("Controller", func() {
+			extensionSecret.Data = gardenSecret.Data
+		})
+
+		It("should create the extension secret and extension BackupBucket if then do not exist yet", func() {
```
Nit: 
```suggestion
		It("should create the extension secret and extension BackupBucket if they do not exist yet", func() {
```

plkokanov commented on pkg/gardenlet/controller/backupentry/reconciler.go:
```diff
@@ -155,7 +157,7 @@ func (r *Reconciler) reconcileBackupEntry(
 	}
 
-	gardenSecret, err := r.getGardenSecret(gardenCtx, backupBucket)
+	gardenCredentials, err := kubernetesutils.GetCredentialsByObjectReference(gardenCtx, r.GardenClient, *backupBucket.Spec.CredentialsRef)
```
When the CredentialRefs points to a `corev1.Secret`  won't there be a value for `backupBucket.Status.GeneratedSecretRef`? And shouldn't we use the data from it instead?

vpnachev replied:
Very nice catch! Backupentry controller indeed must use the GeneratedSecret when it is set in the BackupBucket status. As far as I know, only Azure provider is using this feature and the generated secret is used only for operations related to the "entry", i.e. managing "folders" and "files", and writing data, while operations with the bucket are executed with the credentials from the bb.spec.credetialsRef. 

I will fix it.

plkokanov commented on pkg/gardenlet/controller/backupentry/reconciler.go:
```diff
@@ -321,12 +337,12 @@ func (r *Reconciler) deleteBackupEntry(
 		}
 
-		gardenSecret, err := r.getGardenSecret(gardenCtx, backupBucket)
+		gardenCredentials, err := kubernetesutils.GetCredentialsByObjectReference(gardenCtx, r.GardenClient, *backupBucket.Spec.CredentialsRef)
```
Same as my comment above.

plkokanov reviewed (commented):
Just one question and a minor nit from my side

vpnachev committed:
Use backupBucket.status.generatedSecret for credentials when set

vpnachev committed:
Fix grammar, typo and nit

dimityrmirchev reviewed (approved):
/lgtm

rfranzke reviewed (commented):
/approve

Nice PR number :)
